ooni/scripts/oonideckgen.py

```diff
@@ -0,0 +1,138 @@
+        sys.exit(0)
+
+def generate_deck(options):
+
+    deck_data = {
+        "name": "Default ooniprobe deck",
+        "description": "Default ooniprobe deck generated on {0}".format(
+                       prettyDateNowUTC()),
+        "schedule": "@daily",
+        "tasks": [
+            {
+                "ooni": {
+                    "test_name": "http_invalid_request_line"
+                },
+            },
+            {
+                "ooni": {
+                    "test_name": "http_header_field_manipulation"
+                },
+            },
+            {
+                "ooni": {
+                    "test_name": "web_connectivity",
+                    "file": "$citizenlab_${probe_cc}_urls"
+                },
+            },
+            {
+                "ooni": {
+                    "test_name": "web_connectivity",
+                    "file": "$citizenlab_global_urls"
+                }
+            }
+        ]
+    }
+    if options["collector"] is not None:
+        deck_data["collector"] = options['collector']
+
+    if options["bouncer"] is not None:
+        deck_data["bouncer"] = options['bouncer']
+
+    deck = NGDeck(deck_data=deck_data)
+    with open(options['output'], 'w+') as fw:
+        deck.write(fw)
+
+    print("Deck written to {0}".format(options['output']))
+    print("Run ooniprobe like so:")
+    print("ooniprobe -i {0}".format(options['output']))
+
+
+@defer.inlineCallbacks
+def get_user_country_code():
+    config.privacy.includecountry = True
+    yield probe_ip.lookup()
+    defer.returnValue(probe_ip.geodata['countrycode'])
+
+
+@defer.inlineCallbacks
+def oonideckgen(reactor):
+    options = Options()
+    try:
+        options.parseOptions()
+    except usage.UsageError as error_message:
+        print("%s: %s" % (sys.argv[0], error_message))
+        print(options)
+        sys.exit(1)
+
+    print("Checking for update of resources")
+    yield check_for_update()
+
+    if not options['output']:
+        options['output'] = os.getcwd()
+
+    if not options['country-code']:
+        try:
+            options['country-code'] = yield get_user_country_code()
+        except errors.ProbeIPUnknown:
+            print("Could not determine your IP address.")
+            print("Check your internet connection or specify a country code "
+                  "with -c.")
+            sys.exit(4)
+
+    if len(options['country-code']) != 2:
+        print("%s: --country-code must be 2 characters" % sys.argv[0])
+        sys.exit(2)
+
+    if os.path.isdir(options['output']):
+        options['output'] = os.path.join(options['output'], 'web-full.yaml')
+
+    options['country-code'] = options['country-code'].lower()
+
+    try:
+        os.makedirs(os.path.dirname(options['output']))
+    except OSError as exception:
+        if exception.errno != errno.EEXIST:
+            raise
+
+    generate_deck(options)
+
+def run():
+    task.react(oonideckgen)
```

Comment:
Wow, how would Twisted know that it's time to exit the reactor when it reaches the end of the task? That is, how is the end of the task defined?